google/cloud/forseti/scanner/scanners/retention_scanner.py

```diff
@@ -0,0 +1,141 @@
+    """Pipeline to Bucket acls data from DAO."""
+
+    def __init__(self, global_configs, scanner_configs, service_config,
+                 model_name, snapshot_timestamp, rules):
+        """Initialization.
+
+        Args:
+            global_configs (dict): Global configurations.
+            scanner_configs (dict): Scanner configurations.
+            service_config (ServiceConfig): Forseti 2.0 service configs
+            model_name (str): name of the data model
+            snapshot_timestamp (str): Timestamp, formatted as YYYYMMDDTHHMMSSZ.
+            rules (str): Fully-qualified path and filename of the rules file.
+        """
+        super(RetentionScanner, self).__init__(
+            global_configs,
+            scanner_configs,
+            service_config,
+            model_name,
+            snapshot_timestamp,
+            rules)
+        self.rules_engine = retention_rules_engine.RetentionRulesEngine(
+            rules_file_path=self.rules,
+            snapshot_timestamp=self.snapshot_timestamp)
+        self.rules_engine.build_rule_book(self.global_configs)
+
+    @staticmethod
+    def _flatten_violations(violations):
+        """Flatten RuleViolations into a dict for each RuleViolation member.
+
+        Args:
+            violations (list): The RuleViolations to flatten.
+
+        Yields:
+            dict: Iterator of RuleViolations as a dict per member.
+        """
+        for violation in violations:
+            violation_data = {'describe': violation.violation_describe}
+
+            yield {
+                'resource_name': violation.resource_name,
+                'resource_type': violation.resource_type,
+                'full_name': violation.full_name,
+                'rule_index': violation.rule_index,
+                'rule_name': violation.rule_name,
+                'violation_type': violation.violation_type,
+                'violation_data': violation_data,
+                'resource_data': '',
+                'resource_id': violation.resource_name
+            }
+
+    def _output_results(self, all_violations):
+        """Output results.
+
+        Args:
+            all_violations (list): All violations
+        """
+        all_violations = self._flatten_violations(all_violations)
+        self._output_results_to_db(all_violations)
+
+    def _find_bucket_violations(self, bucket_lifecycle_info):
+        """Find violations in the policies.
+
+        Args:
+            bucket_lifecycle_info (list): Bucket lifecycle to
+                search for violations in.
+
+        Returns:
+            list: All violations.
+        """
+        all_violations = []
+        LOGGER.info('Finding retention violations...')
+
+        for single_info in bucket_lifecycle_info:
+            violations = self.rules_engine.find_buckets_violations(
+                single_info)
+            all_violations.extend(violations)
+
+        return all_violations
+
+    def _retrieve_bucket(self):
```

Comment:
You may be able to use some ideas from https://github.com/GoogleCloudPlatform/forseti-security/pull/2075 to not have to handle specific resources individually